pkg/cloudprovider/aws/amifamily/bootstrap/bottlerocket.go

```diff
@@ -15,59 +15,52 @@ limitations under the License.
 package bootstrap
 
 import (
+	"bytes"
 	"encoding/base64"
 	"fmt"
 
+	"github.com/aws/aws-sdk-go/aws"
 	"github.com/pelletier/go-toml/v2"
 )
 
 type Bottlerocket struct {
 	Options
 }
 
-// config is the root of the bottlerocket config, see more here https://github.com/bottlerocket-os/bottlerocket#using-user-data
-type config struct {
-	Settings settings `toml:"settings"`
-}
-
-// settings is part of the bottlerocket config
-type settings struct {
-	Kubernetes kubernetes `toml:"kubernetes"`
-}
-
-// kubernetes specific configuration for bottlerocket api
-type kubernetes struct {
-	APIServer          string              `toml:"api-server"`
-	ClusterCertificate *string             `toml:"cluster-certificate"`
-	ClusterName        string              `toml:"cluster-name,omitempty"`
-	ClusterDNSIP       string              `toml:"cluster-dns-ip,omitempty"`
-	NodeLabels         map[string]string   `toml:"node-labels,omitempty"`
-	NodeTaints         map[string][]string `toml:"node-taints,omitempty"`
-	MaxPods            int                 `toml:"max-pods,omitempty"`
-}
-
 func (b Bottlerocket) Script() string {
-	s := config{Settings: settings{
-		Kubernetes: kubernetes{
-			ClusterName:        b.ClusterName,
-			APIServer:          b.ClusterEndpoint,
-			ClusterCertificate: b.CABundle,
-			NodeLabels:         b.Labels,
-		},
-	}}
+	s := b.unmarshalCustomUserData()
+	// Karpenter will overwrite settings present inside custom UserData
+	// based on other fields specified in the provisioner
+	s.Settings.Kubernetes.ClusterName = &b.ClusterName
+	s.Settings.Kubernetes.APIServer = b.ClusterEndpoint
+	s.Settings.Kubernetes.ClusterCertificate = b.CABundle
+	s.Settings.Kubernetes.NodeLabels = b.Labels
+
 	if b.KubeletConfig != nil && len(b.KubeletConfig.ClusterDNS) > 0 {
-		s.Settings.Kubernetes.ClusterDNSIP = b.KubeletConfig.ClusterDNS[0]
+		s.Settings.Kubernetes.ClusterDNSIP = &b.KubeletConfig.ClusterDNS[0]
 	}
 	if !b.AWSENILimitedPodDensity {
-		s.Settings.Kubernetes.MaxPods = 110
+		s.Settings.Kubernetes.MaxPods = aws.Int(110)
 	}
 	s.Settings.Kubernetes.NodeTaints = map[string][]string{}
 	for _, taint := range b.Taints {
 		s.Settings.Kubernetes.NodeTaints[taint.Key] = append(s.Settings.Kubernetes.NodeTaints[taint.Key], fmt.Sprintf("%s:%s", taint.Value, taint.Effect))
 	}
-	script, err := toml.Marshal(s)
+	script := new(bytes.Buffer)
+	tomlEncoder := toml.NewEncoder(script)
+	err := tomlEncoder.Encode(&s)
+	if err != nil {
+		panic(err)
+	}
+	return base64.StdEncoding.EncodeToString(script.Bytes())
+}
+
+func (b Bottlerocket) unmarshalCustomUserData() config {
+	var c config
+	b64DecodedBytes, _ := base64.StdEncoding.DecodeString(*b.CustomUserData)
```

Comment:
I saw that it's validated prior, but if `Script()` is changed to possibly return an error, I don't think it hurts to check for an error here instead of ignoring it.

Checking just below here will allow removing an additional panic call.

Reply:
Right yeah, I had left it alone because we do this in our provider's validation code. Changed now.